src/controller/ai_controller.c

```diff
@@ -277,58 +689,117 @@ int ai_block_projectile(controller *ctrl, ctrl_event **ev) {
     return 0;
 }
 
-int ai_controller_poll(controller *ctrl, ctrl_event **ev) {
+/** 
+ * \brief Process the current selected move.
+ *
+ * \param ctrl Controller instance.
+ * \param ev The current controller event.
+ *
+ * \return Void.
+ */
+void process_selected_move(controller *ctrl, ctrl_event **ev) {
     ai *a = ctrl->data;
     object *o = ctrl->har;
-    if (!o) {
-        return 1;
+
+    if(a->input_lag_timer > 0) {
+        a->input_lag_timer--;
+    } else {
+        a->move_str_pos--;
+        if(a->move_str_pos <= 0) {
+            a->move_str_pos = 0;
+        }
+        a->input_lag_timer = a->input_lag;
     }
-    har *h = object_get_userdata(o);
-    object *o_enemy = game_state_get_player(o->gs, h->player_id == 1 ? 0 : 1)->har;
+    int ch = str_at(&a->selected_move->move_string, a->move_str_pos);
+    controller_cmd(ctrl, char_to_act(ch, o->direction), ev);
+}
 
-    // Do not run AI while the game is paused
-    if(game_state_is_paused(o->gs)) { return 0; }
+/** 
+ * \brief Handle the AI's movement.
+ *
+ * \param ctrl Controller instance.
+ * \param ev The current controller event.
+ *
+ * \return Void.
+ */
+void handle_movement(controller *ctrl, ctrl_event **ev) {
+    ai *a = ctrl->data;
+    object *o = ctrl->har;
 
-    // Do not run AI while match is starting or ending
-    // XXX this prevents the AI from doing scrap/destruction moves
-    // XXX this could be fixed by providing a "scene changed" event
-    if(is_arena(game_state_get_scene(o->gs)->id) &&
-       arena_get_state(game_state_get_scene(o->gs)) != ARENA_STATE_FIGHTING) {
+    // Change action after act_timer runs out
+    int jump_thresh = 0;
+    if(a->act_timer <= 0 && rand_int(100) > (90 - (a->difficulty * 3))) {
+        int p_move_roll = rand_int(100);
 
-        // null out selected move to fix the "AI not moving problem"
-        a->selected_move = NULL;
-        return 0;
-    }
+        int p_move_thresh = 16 - (a->difficulty * 2);
 
-    // Grab all projectiles on screen
-    vector_clear(&a->active_projectiles);
-    game_state_get_projectiles(o->gs, &a->active_projectiles);
+        if (p_move_roll > p_move_thresh) {
 
-    // Try to block har
-    if(ai_block_har(ctrl, ev)) {
-        return 0;
+            int p_fwd_roll = rand_int(100);
+
+            int p_fwd_thresh = 50 - ((a->difficulty - 1) * 2);
+            if (a->pilot->pref_fwd >= a->pilot->pref_back) p_fwd_thresh -= 4; // pilot pref
+            if (a->pilot->att_hyper) p_fwd_thresh -= 4; // pilot hyper
+
+            // DEBUG("p_fwd_thresh %d", p_fwd_thresh);
+            
+            if (p_fwd_roll >= p_fwd_thresh) {
+                // walk forward
+                a->cur_act = (o->direction == OBJECT_FACE_RIGHT ? ACT_RIGHT : ACT_LEFT);
+                jump_thresh = 78 - (a->difficulty * 2);
```

Comment:
Comments on where the magic numbers come from would be good to have. They could optionally be #defined constants.

Reply:
So they're currently made up numbers that i have been experimenting with.

I will define them as constants and add some documentation.

Reply:
Addressed in commit: https://github.com/omf2097/openomf/pull/387/commits/9b3d9c1aa6bb99594bdd2bd294eaf6b052645570****